File: solutions/elixir/freelancer-rates/1/lib/freelancer_rates.ex
defmodule FreelancerRates do
  def daily_rate(hourly_rate), do: 8.0 * hourly_rate

  def apply_discount(before_discount, discount), do: before_discount - before_discount * (discount / 100)

  def monthly_rate(hourly_rate, discount) do
    hourly_rate
    |> daily_rate
    |> (fn x -> x * 22 end).()
    |> apply_discount(discount)
    |> ceil
  end

  def days_in_budget(budget, hourly_rate, discount) do
    hourly_rate
    |> daily_rate
    |> apply_discount(discount)
    |> (fn x -> budget / x end).()
    |> Float.floor(1)
  end
end

File: solutions/cpp/freelancer-rates/1/freelancer_rates.cpp
#include <cmath>

double daily_rate(double hourly_rate) {
    return hourly_rate * 8.;
}

double apply_discount(double before_discount, double discount) {
  return before_discount - before_discount * (discount / 100);
}

int monthly_rate(double hourly_rate, double discount) {
  double daily {daily_rate(hourly_rate) * 22.};

  return ceil(apply_discount(daily, discount));
}

int days_in_budget(int budget, double hourly_rate, double discount) {
  double cost {apply_discount(daily_rate(hourly_rate), discount)};
  return floor(budget / cost);
}
